org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/ServiceRuleTest.java

```diff
@@ -0,0 +1,182 @@
+			.requiredWithin(10)
+			.build()) {
+
+			foos.init(getClass());
+		}
+	}
+
+	@Test
+	public void successWhenService() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class).required()
+			.build(); BundleContextRule bcRule = new BundleContextRule()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					100, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			assertThat(foos.isEmpty()).isFalse();
+			assertThat(foos.size()).isEqualTo(1);
+			assertThat(foos.getTimeout()).isEqualTo(ServiceHolder.DEFAULT_TIMEOUT);
+			assertThat(foos.isRequired()).isTrue();
+		}
+	}
+
+	@Test
+	public void successWhenServiceWithTimeout() throws Exception {
+		try (ServiceRule<Foo> foos = new ServiceRule.Builder<>(Foo.class)
+			.requiredWithin(10)
+			.build(); BundleContextRule bcRule = new BundleContextRule()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, null),
+					5, TimeUnit.MILLISECONDS);
+
+			foos.init(getClass());
+
+			assertThat(foos.getService()).isNotNull();
+			assertThat(foos.getServiceReference()).isNotNull();
+			assertThat(foos.getServiceReferences()).isNotNull();
+			assertThat(foos.getServices()).isNotEmpty()
+				.contains(afoo);
+			assertThat(foos.getTimeout()).isEqualTo(10);
+			assertThat(foos.getTracked()).isNotEmpty();
+			assertThat(foos.getTrackingCount()).isEqualTo(1);
+			assertThat(foos.isRequired()).isTrue();
+			assertThat(foos.size()).isEqualTo(1);
+			assertThat(foos.isEmpty()).isFalse();
+			assertThat(foos.waitForService(20)).isNotNull();
+		}
+	}
+
+	@Test
+	public void matchByFilter() throws Exception {
+		try (ServiceRule<Foo> fooRule = new ServiceRule.Builder<>(Foo.class)
+			.required()
+			.filter("(foo=bar)")
+			.build(); BundleContextRule bcRule = new BundleContextRule()) {
+
+			bcRule.init(getClass());
+
+			final Foo afoo = new Foo() {};
+
+			Executors.newSingleThreadScheduledExecutor().schedule(
+					() -> bcRule.getBundleContext()
+						.registerService(Foo.class, afoo, of("foo", "bar")),
+					100, TimeUnit.MILLISECONDS);
+
+			fooRule.init(getClass());
+
+			assertThat(fooRule.getService()).isNotNull();
+			assertThat(fooRule.getServiceReference()).isNotNull();
+			assertThat(fooRule.getServiceReferences()).isNotNull();
+			assertThat(fooRule.getServices()).isNotEmpty()
+				.contains(afoo);
+			assertThat(fooRule.getTimeout()).isEqualTo(ServiceHolder.DEFAULT_TIMEOUT);
+			assertThat(fooRule.getTracked()).isNotEmpty();
+			assertThat(fooRule.getTrackingCount()).isEqualTo(1);
+			assertThat(fooRule.isRequired()).isTrue();
+			assertThat(fooRule.size()).isEqualTo(1);
+			assertThat(fooRule.isEmpty()).isFalse();
```

Comment:
There don't appear to be any tests that test the situation where `getTrackingCount()` and/or `size()` are > 1. Would it be prudent to add some?

Reply:
done